src/i2c.rs

```diff
@@ -7,186 +7,326 @@
 //! extern crate embedded_hal_mock;
 //!
 //! use embedded_hal::prelude::*;
-//! use embedded_hal::blocking::i2c::Read;
-//! use embedded_hal_mock::i2c::Mock as I2cMock;
+//! use embedded_hal::blocking::i2c::{Read, Write, WriteRead};
+//! use embedded_hal_mock::i2c::{Mock as I2cMock, Transaction as I2cTransaction};
 //!
 //! let mut i2c = I2cMock::new();
 //!
-//! // Reading
-//! let mut buf = [0; 3];
-//! i2c.set_read_data(&[1, 2]);
-//! i2c.read(0, &mut buf).unwrap();
-//! assert_eq!(buf, [1, 2, 0]);
-//! assert_eq!(i2c.get_last_address(), Some(0));
+//! // Configure expectations
+//! i2c.expect(vec![
+//!     I2cTransaction::write(0xaa, vec![1u8, 2u8]),
+//!     I2cTransaction::read(0xbb, vec![3u8, 4u8]),
+//! ]);
 //!
 //! // Writing
-//! let buf = [1, 2, 4];
-//! i2c.write(42, &buf).unwrap();
-//! assert_eq!(i2c.get_last_address(), Some(42));
-//! assert_eq!(i2c.get_write_data(), &[1, 2, 4]);
+//! i2c.write(0xaa, &vec![1u8, 2u8]).unwrap();
+//!
+//! // Reading
+//! let mut buf = vec![0u8; 2];
+//! i2c.read(0xbb, &mut buf).unwrap();
+//! assert_eq!(buf, vec![3u8, 4u8]);
+//!
+//! // Finalise expectations
+//! i2c.done();
 //! ```
 
-use std::io::Read;
+use std::collections::VecDeque;
+use std::sync::{Arc, Mutex};
 
 use hal::blocking::i2c;
 
 use error::MockError;
 
-const WRITE_BUF_SIZE: usize = 64;
-
-/// Mock I²C implementation
-///
-/// The main idea behind this is that you set the data that will be read back
-/// by a read call in advance. Then the driver can read from the mock device,
-/// and will get exactly the data you prepared.
-///
-/// See the usage section in the module level docs for an example.
-pub struct Mock<'a> {
-    data: &'a [u8],
-    address: Option<u8>,
-    buf: [u8; WRITE_BUF_SIZE],
-    buf_bytes_written: usize,
+/// I2C Transaction modes
+#[derive(Clone, Debug, PartialEq)]
+pub enum Mode {
+    Write,
+    Read,
+    WriteRead,
+}
+
+/// I2C Transaction type
+/// Models an I2C read or write
+#[derive(Clone, Debug, PartialEq)]
+pub struct Transaction {
+    expected_mode: Mode,
+    expected_addr: u8,
+    expected_data: Vec<u8>,
+    response_data: Vec<u8>,
```

Comment:
I think using an `Option<Vec<u8>>` would be cleaner than creating an empty vector just to denote absence.

Or even this:

```rust
pub enum Mode {
    Write { expected: Vec<u8> },
    Read { response: Vec<u8> },
    WriteRead { expected: Vec<u8>, response: Vec<u8> },
}
```

...which would not allow invalid states to exist at all. (In this case it might be a good idea to rename `Mode` to `Expectation`.)

Reply:
I wrote it that way originally, but, It seemed simpler to check `mode==mode` than to write a match cases to handle and unwrap the right enum combinations, and this way supported the lazy evaluation of expectations when I was trying that.

Reply:
I leave the decision to you :slightly_smiling_face: